### PR TITLE
[SMALLFIX] refactoring master and worker

### PR DESCRIPTION
**Pull request by jsimsa**
This PR:

- mirrors the Alluxio worker class hierarchy (AlluxioWorker, AlluxioWorkerService, DefaultAlluxioWorker) for Alluxio master
- removes System.exit() from DefaultAlluxioWorker constructor (which among other things will prevent hard to debug test failures)

### CODE REVIEW

File: core/server/src/main/java/alluxio/worker/AlluxioWorker.java [hunk 1]
```diff
@@ -38,13 +38,16 @@ public static void main(String[] args) {
       System.exit(-1);
     }
 
-    AlluxioWorkerService worker = new DefaultAlluxioWorker();
+    AlluxioWorkerService worker = null;
     try {
+      worker = new DefaultAlluxioWorker();
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/AlluxioWorker.java]: maybe instantiate this in its own try-catch block. We don't want to call `stop` if the call to `start` failed

File: core/server/src/main/java/alluxio/worker/AlluxioWorker.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * Class for running an Alluxio worker.
+ * Entry point for the Alluxio worker.
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/worker/AlluxioWorker.java]: Could you make the wording of these comments consistent?

File: core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java [hunk 3]
```diff
@@ -102,50 +100,46 @@
   private AtomicReference<Long> mWorkerId;
 
   /**
-   * Constructs a {@link DefaultAlluxioWorker}.
+   * Creates a new instance of {@link DefaultAlluxioWorker}.
+   *
+   * @throws Exception if the creation fails
    */
-  public DefaultAlluxioWorker() {
+  public DefaultAlluxioWorker() throws Exception {
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java]: Either throw typed exceptions (if we plan to handle them) or use Throwables.propagate (if we will not be able to handle them)?

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java [hunk 4]
```diff
@@ -186,358 +74,49 @@ public static void main(String[] args) {
...
-    if (files == null) {
-      return false;
-    }
-    // Search for the format file.
-    String formatFilePrefix = Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX);
-    for (String file : files) {
-      if (file.startsWith(formatFilePrefix)) {
-        return true;
-      }
+      System.exit(-1);
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: Is it possible to abstract this block of logic (repeats in the different services)?

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java [hunk 5]
```diff
@@ -154,7 +42,7 @@ public static void main(String[] args) {
   /**
    * @return the (cached) master service loader
    */
-  private static ServiceLoader<MasterFactory> getServiceLoader() {
+  protected static ServiceLoader<MasterFactory> getServiceLoader() {
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: I feel this and the next static method should be a part of  interface `AlluxioMasterService.Factory` (See my next comment), and leave this class really a simple wrapper

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java [hunk 6]
```diff
@@ -186,358 +74,49 @@ public static void main(String[] args) {
   }
 
   /**
-   * Factory for creating {@link AlluxioMaster} or {@link FaultTolerantAlluxioMaster} based on
-   * {@link Configuration}.
+   * Factory for creating {@link AlluxioMasterService}.
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: is it better to put the factory method create to interface AlluxioMasterService, which seems more like our convention?

File: core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java [hunk 7]
```diff
@@ -102,50 +100,46 @@
...
-        }
+    mStartTimeMs = System.currentTimeMillis();
+    mBlockWorker = new DefaultBlockWorker(mWorkerId);
+    mFileSystemWorker = new DefaultFileSystemWorker(mBlockWorker, mWorkerId);
+
+    mAdditionalWorkers = new ArrayList<>();
+    List<? extends Worker> workers = Lists.newArrayList(mBlockWorker, mFileSystemWorker);
+    // Discover and register the available factories
+    // NOTE: ClassLoader is explicitly specified so we don't need to set ContextClassLoader
+    ServiceLoader<WorkerFactory> discoveredMasterFactories =
```
Comment (apc999) [inline core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java]: `discoveredWorkerFactories`?

File: core/server/src/main/java/alluxio/master/DefaultAlluxioMaster.java [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+  /** The RPC server. */
+  private TServer mMasterServiceServer = null;
+
+  /** is true if the master is serving the RPC server. */
+  private boolean mIsServing = false;
+
+  /** The start time for when the master started serving the RPC server. */
+  private long mStartTimeMs = -1;
+
+  protected DefaultAlluxioMaster() {
```
Comment (apc999) [inline core/server/src/main/java/alluxio/master/DefaultAlluxioMaster.java]: since `DefaultAlluxioWorker` constructor now throws exception, shall we make `DefaultAlluxioMaster` similar?

Comment (apc999) [inline core/server/src/main/java/alluxio/master/DefaultAlluxioMaster.java]: basically, now worker and master can share more code and make it possible to have more consistent behavior. I am fine to leave this as a TODO

File: minicluster/src/main/java/alluxio/master/LocalAlluxioMaster.java [hunk 9]
```diff
@@ -155,13 +155,13 @@ public void clearClients() throws IOException {
...
   public InetSocketAddress getAddress() {
-    return mAlluxioMaster.getMasterAddress();
+    return mAlluxioMaster.getRpcAddress();
   }
 
   /**
    * @return the internal {@link AlluxioMaster}
    */
-  public AlluxioMaster getInternalMaster() {
+  public AlluxioMasterService getInternalMaster() {
```
Comment (apc999) [inline minicluster/src/main/java/alluxio/master/LocalAlluxioMaster.java]: fix javadoc?

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java (hunk 5 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: done

File: core/server/src/main/java/alluxio/master/DefaultAlluxioMaster.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/DefaultAlluxioMaster.java]: done

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java (hunk 4 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: done

File: core/server/src/main/java/alluxio/worker/AlluxioWorker.java (hunk 2 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/AlluxioWorker.java]: done

File: core/server/src/main/java/alluxio/worker/AlluxioWorker.java (hunk 1 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/AlluxioWorker.java]: no longer relevant

File: core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java (hunk 3 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java]: done

File: core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/DefaultAlluxioWorker.java]: done

File: minicluster/src/main/java/alluxio/master/LocalAlluxioMaster.java (hunk 9 above)
Comment (jsimsa) [inline minicluster/src/main/java/alluxio/master/LocalAlluxioMaster.java]: done

**Comment by jsimsa** [comment]
@aaudiber @apc999 @calvinjia thanks for your comments, PTAL

**Comment by jsimsa** [comment]
@apc999 `getMasterAddress` is internal API so changing is backwards compatible

File: core/server/src/main/java/alluxio/master/AlluxioMasterService.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+     * @return the (cached) list of the enabled master services' names
+     */
+    public static synchronized List<String> getServiceNames() {
+      if (sServiceNames != null) {
+        return sServiceNames;
+      }
+      sServiceNames = new ArrayList<>();
+      sServiceNames.add(Constants.BLOCK_MASTER_NAME);
+      sServiceNames.add(Constants.FILE_SYSTEM_MASTER_NAME);
+      sServiceNames.add(Constants.LINEAGE_MASTER_NAME);
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMasterService.java]: Should these be discovered automatically?

Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/AlluxioMasterService.java]: I can make that change (but I would need to create factory classes for BlockMaster, FileSystemMaster, LineageMaster) ... would you like me to do that in this PR?

Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMasterService.java]: Let's do it in a separate PR, it is not really related to this change.

**Comment by calvinjia** [comment]
LGTM
